### PR TITLE
A square-problem feasible point is a solved-band answer, not a solver error

### LINKED ISSUE #815: Local infeasibility reported on a feasible square IDAES flowsheet that Ipopt solves
> [!IMPORTANT]
> **Corrected in [this comment](https://github.com/jkitchin/pounce/issues/815#issuecomment-5451845993).**
> The comparison below is not apples-to-apples: Ipopt was run through
> `writer_config={"linear_presolve": True, "scale_model": True}` while pounce got the
> raw model. The headline finding survives on a corrected, like-for-like comparison
> (`ipopt_v2` converges in 4.5 s, `pounce_v2` fails, both presolved and scaled), but
> the **"Options tried" table and the scaling hypothesis below are superseded** —
> pounce fails even when handed the presolved, scaled problem. Read the comment first.

---

## Summary

On a 536×536 **square** IDAES flowsheet (zero degrees of freedom, trivial `f(x)=0`
objective), pounce exits `Converged to a point of local infeasibility` after
**2–3 iterations**. Ipopt 3.13.2 (the IDAES build) converges on the identical
model from the identical starting point in 7.9 s.

The problem is genuinely feasible, and pounce itself can confirm that: if the
model is first converged with Ipopt and then handed to pounce, pounce terminates
`optimal` in 5.6 s, and its sensitivity backsolves agree with an independent
reduced-Jacobian calculation (PyNumero + ASL Jacobian + sparse solve) to 0.06%.
So the `.nl` is fine and the solution is reachable — pounce cannot get there from
the initialization.

**Model:** an IDAES flowsheet of a naphtha hydrotreater — 5 unit models
(feed/effluent heat exchanger, trim heater, adiabatic CSTR with HDS kinetics,
trim cooler, flash) with a genuine recycle arc, on a Generic property package
with PR-style VLE. 536 equations, 536 unknowns after the specs are fixed. Nothing
exotic; this is the ordinary shape of an IDAES flowsheet.

## Reproducer

```python
m = build_model(); specify_model(m); initialize_model(m)
m._obj = pyo.Objective(expr=0.0)          # square problem needs an objective
pyo.SolverFactory("pounce").solve(m, tee=True)
```

```
### Ipopt ###
CONVERGED in 7.9 s

### pounce 0.10.0 ###
Total number of variables...............: 536
Total number of equality constraints....: 536
Total number of inequality constraints..:   0
   0  0.0000000e+00 5.02e+03 1.00e+00    0.0 0.00e+00      - 0.00e+00 0.00e+00   0
   1  0.0000000e+00 5.02e+03 1.02e+05    3.0 5.88e+06      - 7.26e-01 2.13e-05f  1
   2  0.0000000e+00 4.97e+03 5.09e+05    5.3 3.37e+07      - 1.90e-02 3.49e-03f  1
Number of Iterations....: 2
EXIT: Converged to a point of local infeasibility. Problem may be infeasible.
```

The second-opinion ladder fires and does not recover:

```
pounce: local infeasibility — re-solving along 2 different trajectories before believing it
pounce: feral_scaling=mc64 re-solve did not recover (InfeasibleProblemDetected).
pounce: mu_strategy=adaptive re-solve did not recover (InfeasibleProblemDetected).
pounce: keeping the original local-infeasibility verdict
```

Total wall clock 61.2 s, against Ipopt's 7.9 s.

## What the log suggests

Iteration 0 starts at a constraint violation of 5.02e+03. Accepted steps are tiny
(`2.13e-05f`, `3.49e-03f`, `6.14e-05h`) while the regularization / `du` column is
very large (`5.88e+06`, `3.37e+07`), and pounce concludes after 2–3 iterations.

This has the shape of a scaling problem. IDAES models routinely mix pressures of
~3e6 Pa, molar flows of ~30 mol/s, and trace mole fractions of ~1e-6 in the same
system, and Ipopt's default `gradient-based` NLP scaling is a large part of what
makes them tractable.

## Options tried — all still infeasible

| option | result | time |
|---|---|---|
| defaults | InfeasibleProblemDetected | 60 s |
| `nlp_scaling_method=gradient-based` | InfeasibleProblemDetected | 65 s |
| `bound_relax_factor=0` | InfeasibleProblemDetected | 71 s |
| `mu_init=1e-6` | InfeasibleProblemDetected | 35 s |

## A fairness caveat

Ipopt does not solve this in a single call either. The flowsheet's `solve_model()`
deliberately re-issues `solve()` up to four times, because each call resets the
barrier parameter and restoration state and effectively warm-restarts; a single
Ipopt call with a larger `max_iter` does *not* substitute for that (it wanders
rather than converging).

So part of the gap may simply be that pounce gives up where Ipopt-with-restarts
persists. But pounce stops after 2–3 iterations, far short of any iteration
limit, which is why the infeasibility verdict looks premature rather than
budget-limited.

## Why this matters

This is the standard shape of an IDAES flowsheet, so it likely affects any attempt
to use pounce as a drop-in Ipopt replacement in that ecosystem. It is also a
gating problem for the sensitivity feature specifically: `pyomo_pounce`'s declared
-parameter sensitivity is genuinely nicer than the sIPOPT/k_aug workflow it
replaces (declare, solve, `gradient(...)` as a cheap backsolve — 3.5 ms for four
sensitivities here), but it is only reachable if the solve converges.

**Environment:** pounce 0.10.0 built from `5cf01b16`, macOS arm64 (Apple Silicon),
`pyomo-pounce` from the same tree, Pyomo 6.10.1, IDAES with Ipopt 3.13.2.

### PR DESCRIPTION
**Pull request by jkitchin**
Fixes #815.

`Feasible_Point_Found` wrote AMPL `solve_result_num = 100`. Pyomo's contrib v2
`.sol` reader maps `100..=199` to `TerminationCondition.error`, so a correct
answer reached the caller as a **solver error**; the legacy reader maps the
band to `optimal` with `status=warning` — the same complaint that moved
`Solved_To_Acceptable_Level` out of it in #591. It is now `2`, Ipopt's own
code, in the `0..=99` solved band.

## The justification for `100` had stopped being true of the code

The doc comment defending the divergence said the two statuses do not mean the
same thing: that Ipopt returns `FEASIBLE_POINT_FOUND` only for a square
problem, where a feasible point *is* the solution, while POUNCE used it more
loosely for a usable feasible point that missed the convergence criteria.

POUNCE does not. The status has exactly one producer — `min_c_1nrm.rs`
returning `RestorationOutcome::FeasiblePointFound` — behind exactly one gate in
`resto_inner_solver.rs`, whose first conjunct is `is_square_problem`
(`c.x.dim() == c.y_c.dim()`, a port of
`IpoptCalculatedQuantities::IsSquareProblem`). Same condition as Ipopt, so the
same meaning; and on a square problem the objective is constant, so there is no
remaining criterion the point could be said to have missed.

Two surfaces in this repo had already written the correct reading down and were
being contradicted by that comment: `python/pounce/gams/link.py` maps the status
to `(MODELSTAT_FEASIBLE, SOLVESTAT_NORMAL)` and lists it as a success, and
`issue_390_nonlinear_equality_scale.rs` calls it "a success-band answer — AMPL
`objno` code 2, which every band table reads as SOLVED".

## Changes

| | before | after |
|---|---|---|
| `status_to_solve_result_num` | `100` | `2` |
| `pounce` exit code (non-`-AMPL`) | `1` | `0` |
| `v2._V2_STATUS` | `(unknown, feasible)` | `(convergenceCriteriaSatisfied, optimal)` |
| `sens._STATUS_RESULT` | `(feasible, warning)` | `(optimal, ok)` |

The two `pyomo_pounce` rows are set to exactly what the `.sol` route now
produces for the same solve — the rule both tables' docstrings already state,
and the rule #591 established.

The exit code moves because leaving it would have created a *fresh*
inconsistency: the same process would exit 1 while the `.sol` beside it read
solved.

The five-term gate is extracted verbatim into `square_feasible_point_found` so
the `is_square_problem` conjunct — which is what the whole band argument rests
on — can be held by a test instead of a comment. No logic changed.

## Measured

gh #815 is a 536x536 IDAES naphtha-hydrotreater flowsheet, zero degrees of
freedom. Written the way IDAES writes it for Ipopt
(`writer_config={"scale_model": True}`), through Pyomo end-to-end, residuals in
the model's original units:

| | result | wall | max abs residual | thiophene conversion |
|---|---|---|---|---|
| POUNCE, `scale_model` | `optimal` / `ok` | 1.0 s | 2.208e-06 | 0.39187704 |
| POUNCE, `linear_presolve` + `scale_model` | `optimal` / `ok` | 0.8 s | 2.899e-06 | 0.39187658 |

Before this change those same two solves loaded as `optimal` / **`warning`**
through the legacy reader and as `TerminationCondition.error` through the v2
reader. The `.sol` line moved from `objno 0 100` to `objno 0 2`.

## What this does not fix, and a correction to the issue's premise

The model **as filed** — the raw unscaled `.nl` — is not solvable by either
solver here:

- POUNCE stalls at a constraint violation of ~4.97e+03. An 18-configuration
  option sweep (linear solver, `mu` strategy, scaling method,
  `nlp_scaling_max_gradient`, `mu_init`, `start_with_resto`,
  `expect_infeasible_problem`, bound handling, `mc64`, iteration cap) ended
  `Restoration_Failed` or `Infeasible_Problem_Detected` in every case, at that
  same violation.
- Ipopt 3.13.2/ma27 exits `Restoration Failed` at 2418 iterations from a stall
  at `inf_pr = 1.56e+01`, against POUNCE's 1.60e+01 — the two trajectories are
  bit-identical over iterations 0–3.

What makes the model tractable is IDAES's own per-variable scaling factors
applied by the NL writer. That is a frontend configuration; no
`nlp_scaling_method` substitutes for it. Ipopt did not solve this model in any
configuration tried here, including through the reporter's own script — so the
issue's framing ("Ipopt solves it, POUNCE does not") does not hold, while
POUNCE's actual defect on it was real and is what this PR fixes.

## Tests

- `the_exit_code_and_the_sol_band_never_disagree` — the guard as an `iff` over
  all twenty statuses, so a status added to one channel and not the other fails
  it in either direction. #815 is this predicate failing on
  `FeasiblePointFound`; #591 was the same shape one status over.
- `only_a_square_problem_yields_a_feasible_point_verdict` — squareness alone
  decides, every other input held at a passing value.
- `each_residual_guard_refuses_on_its_own` — the other four conjuncts, each
  refused individually from that passing baseline.
- `a_square_problem_feasible_point_is_in_the_solved_band` — the code and the
  band, with the reason.

**What these are not evidence about.** None of it is end-to-end. No fixture in
the CLI corpus reaches this exit; 72 generated square models (three
nonlinearity families x three start points x three row-scale spreads) under
five option sets failed to reach it as well. The exit needs a model larger than
the corpus carries, and the end-to-end evidence is #815's own 536x536 model
above.

`scripts/sweep-fixtures.sh` moved **0 of 154** fixture-legs. That is a control,
not evidence: the sweep does not read `solve_result_num`, and the only
non-reporting edit here is a verbatim extraction.

## Left split on purpose

Three Python success sets still count the status as a failure —
`_minimize._NLP_SUCCESS_STATUS` (shared by `_curve_fit`), `jax._path._OK_STATUS`
and `torch._path._OK_STATUS`. Unlike the CLI they are a scipy-style `success`
flag on a library call that writes no `.sol`, so they were not put in
contradiction with a file the same process had just written, and widening a
public return value deserves its own decision. Filed as #820 so it has an owner
rather than reading as an oversight.

#817, opened earlier against this issue, is **not** a fix for it and is not
part of this PR.

### CODE REVIEW

**Comment by jkitchin** [comment]
Folding this into #817 — the two changes came out of the same investigation and
belong in one branch. The commit is cherry-picked there unchanged
(`fe72d3f5` → `ec85a88b`), along with the fixes for #819 and #820, which are
also gh #815 work rather than separate follow-ups.

Closing unmerged; nothing here is dropped.
